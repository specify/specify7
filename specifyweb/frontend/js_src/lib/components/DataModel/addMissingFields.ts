import { parserFromType } from '../../utils/parser/definitions';
import type { DeepPartial, RA } from '../../utils/types';
import { filterArray, setDevelopmentGlobal } from '../../utils/types';
import { formatUrl } from '../Router/queryString';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type { AnySchema, SerializedResource } from './helperTypes';
import { getScopingResource } from './scoping';
import type { LiteralField, Relationship } from './specifyField';
import { strictGetTable } from './tables';
import type { Tables } from './types';

type ResourceSpec = {
  readonly requiredFields: 'define' | 'omit' | 'set';
  readonly optionalFields: 'define' | 'omit' | 'set';
  readonly toManyRelationships: 'define' | 'omit' | 'set';
  readonly requiredRelationships: 'define' | 'omit' | 'set';
  readonly optionalRelationships: 'define' | 'omit' | 'set';
};

/**
 * This function can:
 * Set missing required fields to literals.
 * Set missing optional fields to null
 * Set missing -to-many relationships to null
 * Set missing dependent -to-one relationships to new objects
 * Do all of these recursively
 */
export function addMissingFields<TABLE_NAME extends keyof Tables>(
  tableName: TABLE_NAME,
  record: DeepPartial<SerializedResource<Tables[TABLE_NAME]>>,
  {
    requiredFields = 'set',
    optionalFields = 'define',
    toManyRelationships = 'set',
    requiredRelationships = 'set',
    optionalRelationships = 'define',
  }: Partial<ResourceSpec> = {}
): SerializedResource<Tables[TABLE_NAME]> {
  const table = strictGetTable(tableName);
  const spec = {
    requiredFields,
    optionalFields,
    toManyRelationships,
    requiredRelationships,
    optionalRelationships,
  };

  const scoping = getScopingResource(table);

  return {
    // This is needed to preserve unknown fields
    ...record,
    ...(Object.fromEntries(
      filterArray(
        table.fields.map((field) =>
          shouldIncludeField(field, spec, record.id === undefined)
            ? [
                field.name,
                field.isRelationship
                  ? handleRelationship(record, field, spec)
                  : record[field.name as keyof typeof record] ??
                    (field.name === 'version'
                      ? 1
                      : (
                          field.isRequired
                            ? requiredFields === 'set'
                            : optionalFields === 'set'
                        )
                      ? parserFromType(field.type).value
                      : null),
              ]
            : undefined
        )
      )
    ) as SerializedResource<Tables[TABLE_NAME]>),
    ...(scoping === undefined
      ? undefined
      : {
          [scoping.relationship.name]:
            record[scoping.relationship.name as 'id'] ??
            (typeof record.id === 'number' ? undefined : scoping.resourceUrl) ??
            null,
        }),
    /*
     * REFACTOR: convert all usages of this to camel case
     */
    resource_uri: record.resource_uri,
    // REFACTOR: consider replacing this with a symbol
    _tableName: tableName,
  };
}
setDevelopmentGlobal('_addMissingFields', addMissingFields);

function shouldIncludeField(
  field: LiteralField | Relationship,
  {
    requiredFields,
    optionalFields,
    toManyRelationships,
    requiredRelationships,
    optionalRelationships,
  }: ResourceSpec,
  isNew: boolean
): boolean {
  if (field.isRelationship) {
    if (relationshipIsToMany(field)) {
      return !field.isDependent() && isNew
        ? false
        : toManyRelationships !== 'omit' && field.type !== 'many-to-many';
    } else
      return (
        (field.isRequired ? requiredRelationships : optionalRelationships) !==
        'omit'
      );
  } else return (field.isRequired ? requiredFields : optionalFields) !== 'omit';
}

function handleRelationship<TABLE_NAME extends keyof Tables>(
  record: DeepPartial<SerializedResource<Tables[TABLE_NAME]>>,
  field: Relationship,
  spec: ResourceSpec
) {
  if (relationshipIsToMany(field))
    if (field.isDependent()) {
      const records = record[field.name as keyof typeof record] as
        | RA<DeepPartial<SerializedResource<AnySchema>>>
        | undefined;
      return (
        records?.map((record) =>
          addMissingFields(field.relatedTable.name, record, spec)
        ) ?? (spec.toManyRelationships === 'set' ? [] : null)
      );
    } else {
      const otherSideName = field.getReverse()?.name;
      return (
        record[field.name as keyof Tables[TABLE_NAME]['toManyIndependent']] ??
        (typeof otherSideName === 'string' && typeof record.id === 'number'
          ? formatUrl(`/api/specify/${field.relatedTable.name}`, {
              [otherSideName]: record.id,
            })
          : undefined)
      );
    }
  else {
    const shouldSet = field.isRequired
      ? spec.requiredRelationships === 'set'
      : spec.optionalRelationships === 'set';
    return (
      record[field.name as keyof typeof record] ??
      (field.isDependent() && shouldSet
        ? addMissingFields(
            field.relatedTable.name,
            (record[field.name as keyof typeof record] as Partial<
              SerializedResource<AnySchema>
            >) ?? {},
            spec
          )
        : null)
    );
  }
}
