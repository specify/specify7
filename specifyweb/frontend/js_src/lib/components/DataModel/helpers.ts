import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
