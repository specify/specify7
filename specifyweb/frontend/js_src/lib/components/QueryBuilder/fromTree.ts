import { queryText } from '../../localization/query';
import type { IR, RA, RR } from '../../utils/types';
import { defined } from '../../utils/types';
import type { AnyTree, SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { getDomainResource } from '../DataModel/scoping';
import { getTreeTable, tables } from '../DataModel/tables';
import type {
  SpQuery,
  SpQueryField,
  Tables,
  TaxonTreeDefItem,
} from '../DataModel/types';
import { softFail } from '../Errors/Crash';
import { hasTablePermission } from '../Permissions/helpers';
import { formatTreeRank } from '../WbPlanView/mappingHelpers';
import { queryFieldFilters } from './FieldFilter';
import { QueryFieldSpec } from './fieldSpec';
import { flippedSortTypes } from './helpers';
import { createQuery } from './index';

export function makeQueryField(
  tableName: keyof Tables,
  path: string,
  options: Partial<SerializedResource<SpQueryField>>
): SpecifyResource<SpQueryField> {
  const field = QueryFieldSpec.fromPath(tableName, path.split('.'))
    .toSpQueryField()
    .set('sortType', flippedSortTypes.none);

  Object.entries(options).forEach(([key, value]) =>
    field.set(
      key as keyof SpQueryField['fields'],
      (value as string | undefined) ?? null
    )
  );

  return field;
}

const makeField = (
  path: string,
  options: Partial<SerializedResource<SpQueryField>>
): SpecifyResource<SpQueryField> =>
  makeQueryField('CollectionObject', path, options);

const defaultFields: RR<
  AnyTree['tableName'],
  (
    nodeId: number,
    rankName: string
  ) => Promise<RA<SpecifyResource<SpQueryField>>>
> = {
  Taxon: async (nodeId, rankName) => [
    makeField('catalogNumber', {
      sortType: flippedSortTypes.ascending,
    }),
    makeField('determinations.taxon.fullName', {}),
    makeField(`determinations.taxon.${rankName}.taxonId`, {
      operStart: queryFieldFilters.equal.id,
      startValue: nodeId.toString(),
      isDisplay: false,
    }),
    makeField('determinations.isCurrent', {
      operStart: queryFieldFilters.trueOrNull.id,
      isDisplay: false,
    }),
    makeField('collectingEvent.locality.localityName', {}),
  ],
  Geography: async (nodeId, rankName) => [
    makeField('catalogNumber', {}),
    makeField('determinations.taxon.fullName', {
      sortType: flippedSortTypes.ascending,
    }),
    makeField('determinations.isCurrent', {
      isDisplay: false,
      operStart: queryFieldFilters.trueOrNull.id,
    }),
    makeField('collectingEvent.locality.localityName', {}),
    makeField('collectingEvent.locality.geography.fullName', {
      sortType: flippedSortTypes.ascending,
    }),
    makeField(`collectingEvent.locality.geography.${rankName}.geographyId`, {
      isDisplay: false,
      operStart: queryFieldFilters.equal.id,
      startValue: nodeId.toString(),
    }),
  ],
  Storage: async (nodeId, rankId) => [
    makeField('catalogNumber', {
      sortType: flippedSortTypes.ascending,
    }),
    makeField('determinations.taxon.fullName', {}),
    makeField('determinations.isCurrent', {
      isDisplay: false,
      operStart: queryFieldFilters.trueOrNull.id,
    }),
    makeField('preparations.storage.fullName', {}),
    makeField(`preparations.storage.${rankId}.storageId`, {
      isDisplay: false,
      operStart: queryFieldFilters.equal.id,
      startValue: nodeId.toString(),
    }),
  ],
  async GeologicTimePeriod(nodeId, rankName) {
    const paleoPath = await fetchPaleoPath();
    return [
      makeField('catalogNumber', {}),
      makeField('determinations.taxon.fullName', {
        sortType: flippedSortTypes.ascending,
      }),
      makeField('determinations.isCurrent', {
        isDisplay: false,
        operStart: queryFieldFilters.trueOrNull.id,
      }),

      ...(typeof paleoPath === 'string'
        ? [
            makeField(`${paleoPath}.chronosStrat.fullName`, {}),
            makeField(
              `${paleoPath}.chronosStrat.${rankName}.geologicTimePeriodId`,
              {
                isDisplay: false,
                operStart: queryFieldFilters.equal.id,
                startValue: nodeId.toString(),
              }
            ),
          ]
        : []),
    ];
  },
  async LithoStrat(nodeId, rankName) {
    const paleoPath = await fetchPaleoPath();
    return [
      makeField('catalogNumber', {}),
      makeField('determinations.taxon.fullName', {
        sortType: flippedSortTypes.ascending,
      }),
      makeField('determinations.isCurrent', {
        isDisplay: false,
        operStart: queryFieldFilters.trueOrNull.id,
      }),
      ...(typeof paleoPath === 'string'
        ? [
            makeField(`${paleoPath}.lithoStrat.fullName`, {}),
            makeField(`${paleoPath}.lithoStrat.${rankName}.lithoStratId`, {
              operStart: queryFieldFilters.equal.id,
              startValue: nodeId.toString(),
              isDisplay: false,
            }),
          ]
        : []),
    ];
  },
};

async function fetchPaleoPath(): Promise<string | undefined> {
  if (!hasTablePermission('Discipline', 'read')) return undefined;
  const paths: IR<string> = {
    collectionobject: 'paleoContext',
    collectingevent: 'collectingevent.paleoContext',
    locality: 'collectingevent.locality.paleoContext',
  };
  const paleoContextTable =
    (await getDomainResource('discipline')?.fetch())?.get(
      'paleoContextChildTable'
    ) ?? '';
  const paleoPath = paths[paleoContextTable];
  if (paleoPath === undefined)
    softFail(new Error('unknown paleoContext child table'));
  return paleoPath;
}

/**
 * Generate a Query for fetchign Collection Objects associated with a given
 * tree node
 */
export async function queryFromTree(
  tableName: AnyTree['tableName'],
  nodeId: number
): Promise<SpecifyResource<SpQuery>> {
  const tree = defined(
    getTreeTable(tableName),
    `Unable to contract a tree query from the ${tableName} table`
  );
  const node = new tree.Resource({ id: nodeId }, { noBusinessRules: true });
  await node.fetch();

  const query = createQuery(
    queryText.treeQueryName({
      tableName: tables.CollectionObject.label,
      nodeFullName: node.get('fullName') ?? node.get('name'),
    }),
    tables.CollectionObject
  );

  const rank: SpecifyResource<TaxonTreeDefItem> = await node.rgetPromise(
    'definitionItem'
  );
  query.set(
    'fields',
    await defaultFields[tree.name](
      nodeId,
      formatTreeRank(rank.get('name') ?? rank.get('title'))
    )
  );

  return query;
}
