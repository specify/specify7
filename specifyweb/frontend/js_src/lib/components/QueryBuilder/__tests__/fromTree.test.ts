import { overrideAjax } from '../../../tests/ajax';
import { requireContext } from '../../../tests/helpers';
import { addMissingFields } from '../../DataModel/addMissingFields';
import { getResourceApiUrl } from '../../DataModel/resource';
import { allTrees } from '../../InitialContext/treeRanks';
import { queryFromTree } from '../fromTree';

requireContext();

const rankNames = {
  Taxon: 'Species',
  Geography: 'County',
  Storage: 'Cabinet',
  GeologicTimePeriod: 'Epoch',
  LithoStrat: 'Formation',
};
const fullNames = {
  Taxon: 'Carpiodes velifer',
  Geography: 'Los Angeles County',
  Storage: 'Cabinet 1',
  GeologicTimePeriod: 'Paleocene',
  LithoStrat: 'Cretaceous',
};

// TODO: Add static data for Tectonic Unit and replace ALL testingTrees usages with allTrees
export const testingTrees = allTrees.filter((t) => t !== 'TectonicUnit');
testingTrees.map((table, rankId) => {
  const nodeId = rankId * 4;
  const rankUrl = getResourceApiUrl(`${table}TreeDefItem`, rankId);
  overrideAjax(getResourceApiUrl(table, nodeId), () =>
    addMissingFields(table, {
      id: nodeId,
      definitionItem: rankUrl,
      // @ts-expect-error Remove error suppress after adding static data for TectonicUnit
      fullName: fullNames[table],
      resource_uri: getResourceApiUrl(table, nodeId),
    })
  );
  overrideAjax(rankUrl, () =>
    addMissingFields(`${table}TreeDefItem`, {
      id: rankId,
      // @ts-expect-error Remove error suppress after adding static data for TectonicUnit
      name: rankNames[table],
      resource_uri: rankUrl,
    })
  );
});

test('queryFromTree', async () =>
  expect(
    Promise.all(
      // TODO: Add static data for Tectonic Unit and replace testingTrees with allTrees
      testingTrees.map(async (tree, index) => queryFromTree(tree, index * 4))
    )
  ).resolves.toMatchSnapshot());
