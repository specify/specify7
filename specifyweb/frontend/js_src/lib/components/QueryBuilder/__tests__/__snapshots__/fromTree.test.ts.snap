// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queryFromTree 1`] = `
[
  {
    "_tableName": "SpQuery",
    "contextname": "CollectionObject",
    "contexttableid": 1,
    "countonly": false,
    "fields": [
      {
        "_tableName": "SpQueryField",
        "fieldname": "catalogNumber",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1.collectionobject.catalogNumber",
        "tablelist": "1",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 1,
        "startvalue": "",
        "stringid": "1,9-determinations,4.taxon.fullName",
        "tablelist": "1,9-determinations,4",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "isCurrent",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 13,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations.determination.isCurrent",
        "tablelist": "1,9-determinations",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "localityName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,10,2.locality.localityName",
        "tablelist": "1,10,2",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 1,
        "startvalue": "",
        "stringid": "1,10,2,3.geography.fullName",
        "tablelist": "1,10,2,3",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "County ID",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 1,
        "sorttype": 0,
        "startvalue": "0",
        "stringid": "1,10,2,3.geography.County ID",
        "tablelist": "1,10,2,3",
      },
    ],
    "formatauditrecids": false,
    "isfavorite": true,
    "name": "Collection Object using \\"Los Angeles County\\"",
    "ordinal": 32767,
    "selectdistinct": false,
    "smushed": false,
    "specifyuser": "/api/specify/specifyuser/2/",
  },
  {
    "_tableName": "SpQuery",
    "contextname": "CollectionObject",
    "contexttableid": 1,
    "countonly": false,
    "fields": [
      {
        "_tableName": "SpQueryField",
        "fieldname": "catalogNumber",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 1,
        "startvalue": "",
        "stringid": "1.collectionobject.catalogNumber",
        "tablelist": "1",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations,4.taxon.fullName",
        "tablelist": "1,9-determinations,4",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "isCurrent",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 13,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations.determination.isCurrent",
        "tablelist": "1,9-determinations",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,63-preparations,58.storage.fullName",
        "tablelist": "1,63-preparations,58",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "Cabinet ID",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 1,
        "sorttype": 0,
        "startvalue": "4",
        "stringid": "1,63-preparations,58.storage.Cabinet ID",
        "tablelist": "1,63-preparations,58",
      },
    ],
    "formatauditrecids": false,
    "isfavorite": true,
    "name": "Collection Object using \\"Cabinet 1\\"",
    "ordinal": 32767,
    "selectdistinct": false,
    "smushed": false,
    "specifyuser": "/api/specify/specifyuser/2/",
  },
  {
    "_tableName": "SpQuery",
    "contextname": "CollectionObject",
    "contexttableid": 1,
    "countonly": false,
    "fields": [
      {
        "_tableName": "SpQueryField",
        "fieldname": "catalogNumber",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 1,
        "startvalue": "",
        "stringid": "1.collectionobject.catalogNumber",
        "tablelist": "1",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations,4.taxon.fullName",
        "tablelist": "1,9-determinations,4",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "Species ID",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 1,
        "sorttype": 0,
        "startvalue": "8",
        "stringid": "1,9-determinations,4.taxon.Species ID",
        "tablelist": "1,9-determinations,4",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "isCurrent",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 13,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations.determination.isCurrent",
        "tablelist": "1,9-determinations",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "localityName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,10,2.locality.localityName",
        "tablelist": "1,10,2",
      },
    ],
    "formatauditrecids": false,
    "isfavorite": true,
    "name": "Collection Object using \\"Carpiodes velifer\\"",
    "ordinal": 32767,
    "selectdistinct": false,
    "smushed": false,
    "specifyuser": "/api/specify/specifyuser/2/",
  },
  {
    "_tableName": "SpQuery",
    "contextname": "CollectionObject",
    "contexttableid": 1,
    "countonly": false,
    "fields": [
      {
        "_tableName": "SpQueryField",
        "fieldname": "catalogNumber",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1.collectionobject.catalogNumber",
        "tablelist": "1",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 1,
        "startvalue": "",
        "stringid": "1,9-determinations,4.taxon.fullName",
        "tablelist": "1,9-determinations,4",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "isCurrent",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 13,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations.determination.isCurrent",
        "tablelist": "1,9-determinations",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,32,46-chronosStrat.geologictimeperiod.fullName",
        "tablelist": "1,32,46-chronosStrat",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "Epoch ID",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 1,
        "sorttype": 0,
        "startvalue": "12",
        "stringid": "1,32,46-chronosStrat.geologictimeperiod.Epoch ID",
        "tablelist": "1,32,46-chronosStrat",
      },
    ],
    "formatauditrecids": false,
    "isfavorite": true,
    "name": "Collection Object using \\"Paleocene\\"",
    "ordinal": 32767,
    "selectdistinct": false,
    "smushed": false,
    "specifyuser": "/api/specify/specifyuser/2/",
  },
  {
    "_tableName": "SpQuery",
    "contextname": "CollectionObject",
    "contexttableid": 1,
    "countonly": false,
    "fields": [
      {
        "_tableName": "SpQueryField",
        "fieldname": "catalogNumber",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1.collectionobject.catalogNumber",
        "tablelist": "1",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 1,
        "startvalue": "",
        "stringid": "1,9-determinations,4.taxon.fullName",
        "tablelist": "1,9-determinations,4",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "isCurrent",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 13,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations.determination.isCurrent",
        "tablelist": "1,9-determinations",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,32,100.lithostrat.fullName",
        "tablelist": "1,32,100",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "Formation ID",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 1,
        "sorttype": 0,
        "startvalue": "16",
        "stringid": "1,32,100.lithostrat.Formation ID",
        "tablelist": "1,32,100",
      },
    ],
    "formatauditrecids": false,
    "isfavorite": true,
    "name": "Collection Object using \\"Cretaceous\\"",
    "ordinal": 32767,
    "selectdistinct": false,
    "smushed": false,
    "specifyuser": "/api/specify/specifyuser/2/",
  },
  {
    "_tableName": "SpQuery",
    "contextname": "CollectionObject",
    "contexttableid": 1,
    "countonly": false,
    "fields": [
      {
        "_tableName": "SpQueryField",
        "fieldname": "catalogNumber",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1.collectionobject.catalogNumber",
        "tablelist": "1",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 1,
        "startvalue": "",
        "stringid": "1,9-determinations,4.taxon.fullName",
        "tablelist": "1,9-determinations,4",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "isCurrent",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 13,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,9-determinations.determination.isCurrent",
        "tablelist": "1,9-determinations",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "fullName",
        "isdisplay": true,
        "isnot": false,
        "isrelfld": false,
        "operstart": 8,
        "sorttype": 0,
        "startvalue": "",
        "stringid": "1,32,1027.tectonicunit.fullName",
        "tablelist": "1,32,1027",
      },
      {
        "_tableName": "SpQueryField",
        "fieldname": "Substructure ID",
        "isdisplay": false,
        "isnot": false,
        "isrelfld": false,
        "operstart": 1,
        "sorttype": 0,
        "startvalue": "20",
        "stringid": "1,32,1027.tectonicunit.Substructure ID",
        "tablelist": "1,32,1027",
      },
    ],
    "formatauditrecids": false,
    "isfavorite": true,
    "name": "Collection Object using \\"Plate\\"",
    "ordinal": 32767,
    "selectdistinct": false,
    "smushed": false,
    "specifyuser": "/api/specify/specifyuser/2/",
  },
]
`;
